woltka/coverage.py

```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/env python3
+
+# ----------------------------------------------------------------------------
+# Copyright (c) 2020--, Qiyun Zhu.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+
+"""Functions for calculating coverage of subject sequences.
+
+The "coverage" refers to the ranges of a subject sequence (e.g., genome) that
+are covered by at least one query sequence (e.g., short sequencing read).
+"""
+
+from os import makedirs
+from os.path import join
+from collections import deque
+
+
+def merge_ranges(ranges):
+    """Merge short, fragmental ranges into long, continuous ones.
+
+    Parameters
+    ----------
+    ranges : list of (int, int)
+        Ranges (start, end) to merge.
+
+    Returns
+    -------
+    list of (int, int)
+        Merged ranges.
+    
+    Notes
+    -----
+    Ranges that have overlaps will be merged into one. For example:
+
+    >>> merge_ranges([(1, 3), (2, 4), (6, 8), (7, 9)])
+    [(1, 4), (6, 9)]
+    """
+    res = []
+    res_append = res.append
```

Comment:
What is the purpose of this line?  Is this an optimization?  Why not just directly call res.append in the rest of this method?

Reply:
Yes this is an optimization. It stores a reference to the method call so that Python doesn't need to search the attribute space of the variable for the particular method every time.